Differential tarball tool

ilanschnell opened the pull request:
This PR adds the `conda-diff-tar` command to conda-mirror, a tool for creating tarballs from a previous reference point of a conda repository.  The file `diff-tar.md` explains all the details.

ilanschnell commented:
Ouch, I should probably add some tests.

scopatz commented on conda_mirror/diff_tar.py:
```diff
@@ -0,0 +1,191 @@
+import os
```
Can you please add a module-level docstirng

ilanschnell replied:
Just did.

scopatz commented on conda_mirror/diff_tar.py:
```diff
@@ -0,0 +1,191 @@
+
+
+mirror_dir = None
+reference_path = './reference.json'
```
global constants should be UPPERCASE :wink:

ilanschnell replied:
Just did

scopatz commented on conda_mirror/diff_tar.py:
```diff
@@ -0,0 +1,191 @@
+        with open(reference_path) as fi:
+            return json.load(fi)
+    except FileNotFoundError:
+        sys.exit('No such file: %s' % reference_path)
```
You probably shouldn't call `sys.exit` from a function like this.  Just raise a system exit or similar

ilanschnell replied:
Fixed

scopatz commented on conda_mirror/diff_tar.py:
```diff
@@ -0,0 +1,191 @@
+        t.add(join(mirror_dir, f), f)
+    t.close()
+    if verbose:
+        print("Wrote: %s" % outfile)
```
It is OK to use f-strings :wink:

ilanschnell replied:
I'm also a C developer ;)

scopatz replied:
haha, so I am I, and yet...

scopatz commented on conda_mirror/diff_tar.py:
```diff
@@ -0,0 +1,191 @@
+
+
+def main():
+    from optparse import OptionParser
```
Woah, I thought this was deprecated in favor of argparse

ilanschnell replied:
Oh, I guess it is deprecated now.  It's just what I've always been using.

scopatz replied:
Haha That's OK!

ilanschnell replied:
Switched to argparse 843b403a5f46ee815dccc7fd0ad5451867b2d643

scopatz commented on setup.py:
```diff
@@ -27,7 +27,8 @@
         "console_scripts": [
-            'conda-mirror = conda_mirror.conda_mirror:cli'
+            'conda-mirror = conda_mirror.conda_mirror:cli',
+            'conda-diff-tar = conda_mirror.diff_tar:main',
```
Shouldn't this be a subcommand of conda-mirror, rather than a separate command?

ilanschnell replied:
Not sure, I realized that just adding the new functionality as options wouldn't make too much sense.  Having a sub command it of course possible too, for example `conda-mirror diff-tar`.

scopatz replied:
@xhochy @mariusvniekerk - any opinions?

scopatz commented on test/test_diff_tar.py:
```diff
@@ -0,0 +1,106 @@
+import os
```
The rest of the test suite uses pytest.  Please adjust this to confom to that style. Thanks!

ilanschnell replied:
Hmm, pytest supports unittest out of the box, see https://docs.pytest.org/en/latest/unittest.html .  I can run `pytest test_diff_tar.py` just fine, so I'm not sure what to adjust here.

xhochy replied:
While they execute just fine, it is a totally different style which makes maintenance harder. Please change the tests here to use pytest-style with fixtures.

ilanschnell commented:
Thanks for the feedback @scopatz.  I've addressed the points.  The only remaining question is how the new functionality should be invoked.  First, I though that adding options to the existing `conda-mirror` command would be best, but then I realized that the new functionality is really for a very special (and not very common) case and adding the options to the existing command would probably only be confusing, in particular since most users are never going to need the functionality.  This is why I added the `conda-diff-tar` command.  Another option would be to add a sub-command `conda-mirror diff-tar`, but since `conda-mirror` doesn't have any sub-commands so far, this would be awkward too.  Thoughts?

xhochy commented on conda_mirror/diff_tar.py:
```diff
@@ -0,0 +1,209 @@
+    if not args.repo_dir:
+        p.error('exactly one REPOSITORY is required, try -h')
+
+    global MIRROR_DIR
```
Please don't use global variables but pass this as argument to functions.

ilanschnell replied:
I don't think there is anything wrong with using a global here.  It makes the code shorter and easier to read, because otherwise you have to pass the variable through several functions.

scopatz replied:
I think that @xhochy that is better because it is encapsulated

ilanschnell replied:
Yes, more encapsulated but also more code.  Anyway, I switched to using pytest fixtures which is really nice because it is less verbose.  Thanks @xhochy for the suggestion.

ilanschnell replied:
I removed the use use of `global MIRROR_DIR` in favor of passing the directory location to as a function argument.

xhochy commented on conda_mirror/diff_tar.py:
```diff
@@ -0,0 +1,209 @@
+
+    if args.create:
+        tar_repo(verbose=args.verbose)
+        return
```
Instead of empty returns, please use `elif` for the following if-statements.

ilanschnell replied:
Thanks, good idea 40a8597955fcea4415d7fd170856283f6afc4e28

xhochy commented on conda_mirror/diff_tar.py:
```diff
@@ -0,0 +1,209 @@
+    tarball.
+    """
+    try:
+        d1 = read_reference()
```
Please `os.exists` instead of exception-based programming. You can raise the `sys.exit` directly in `read_reference`

ilanschnell replied:
I guess you mean `os.path.exists`.  I'll call `sys.exit` in the function directly again, based on the exception because "asking for forgiveness is better than asking for permission" :-)
b260c5bbd65d7408e46f95882434807f774f8c48

scopatz commented on conda_mirror/diff_tar.py:
```diff
@@ -0,0 +1,201 @@
+        sys.exit("""\
+Error: no such file: %s
+Please use the --reference option before creating a differential tarball.
+""" % REFERENCE_PATH)
```
Please do not use sys.exit()

xhochy replied:
@scopatz What else do you suggest? Raise an error?

scopatz replied:
Yeah, just raise an error.  The top-level main function can handle pretty printing errors, if desired.  Library functions should never call sys.exit

ilanschnell commented:
Thanks @scopatz for suggesting to raise an error and handling it in main.  I've made that change, and also added a test for the error being raised in `read_reference()` when the reference file is not present.  I hope we are ready to merge this now.

xhochy commented:
Thanks @ilanschnell !

scopatz commented:
I am going to cut a release with this capability
